From 6a8edb3e82f37904981e7d56fe8a2a065eddd9d6 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 2 May 2014 14:37:10 -0400
Subject: [misc] mei: clear write cb from waiting list on reset

Message-id: <1399041470-21330-38-git-send-email-prarit@redhat.com>
Patchwork-id: 79328
O-Subject: [RHEL7.1 PATCH BZ 1088953 37/77] mei: clear write cb from waiting list on reset
Bugzilla: 1088953
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1088953

commit 30c54df7cb9b15b222529a028390b9c9582dd65e
Author: Alexander Usyskin <alexander.usyskin@intel.com>
Date:   Mon Jan 27 22:27:23 2014 +0200

    mei: clear write cb from waiting list on reset

    Clear write callbacks sitting in write_waiting list on reset.
    Otherwise these callbacks are left dangling and cause memory leak.

    Signed-off-by: Alexander Usyskin <alexander.usyskin@intel.com>
    Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
    Cc: stable <stable@vger.kernel.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Cc: Dean Nelson <dnelson@redhat.com>
Cc: Stefan Assmann <sassmann@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/misc/mei/client.c b/drivers/misc/mei/client.c
index 1ee2b94..ccdacb5 100644
--- a/drivers/misc/mei/client.c
+++ b/drivers/misc/mei/client.c
@@ -942,8 +942,16 @@ void mei_cl_all_wakeup(struct mei_device *dev)
 void mei_cl_all_write_clear(struct mei_device *dev)
 {
  struct mei_cl_cb *cb, *next;
+ struct list_head *list;
 
- list_for_each_entry_safe(cb, next, &dev->write_list.list, list) {
+ list = &dev->write_list.list;
+ list_for_each_entry_safe(cb, next, list, list) {
+  list_del(&cb->list);
+  mei_io_cb_free(cb);
+ }
+
+ list = &dev->write_waiting_list.list;
+ list_for_each_entry_safe(cb, next, list, list) {
   list_del(&cb->list);
   mei_io_cb_free(cb);
  }
-- 
1.7.1